agera/src/main/java/com/google/android/agera/FunctionCompiler.java

```diff
@@ -32,16 +32,46 @@
 
 @SuppressWarnings({"unchecked, rawtypes"})
 final class FunctionCompiler implements FList, FItem {
+  private static final ThreadLocal<FunctionCompiler> compilers = new ThreadLocal<>();
+
+  @NonNull
+  static FunctionCompiler functionCompiler() {
+    FunctionCompiler compiler = compilers.get();
+    if (compiler == null) {
+      compiler = new FunctionCompiler();
+    } else {
+      // Remove compiler from the ThreadLocal to prevent reuse in the middle of a compilation.
+      // recycle(), called by compile(), will return the compiler here. ThreadLocal.set(null) keeps
+      // the entry (with a null value) whereas remove() removes the entry; because we expect the
+      // return of the compiler, don't use the heavier remove().
+      compilers.set(null);
+    }
+    return compiler;
+  }
+
+  @NonNull
+  private static void recycle(@NonNull final FunctionCompiler compiler) {
+    compiler.directives.clear();
+    compilers.set(compiler);
+  }
+
+  private static final int APPLY = 0;
```

Comment:
Don't see the point though. valueOf will be executed hitting the cache, no creation of new objects. Where it really matters (in the loop below) we have to go from Integer -> int anyway, and there the extra set of constants won't help us. Leaving it as it is